In [2]:
import pandas as pd

In [3]:
test_2020 = pd.read_csv('../know_Data/test/KNOW_2020_test.csv')

In [4]:
test_2020

idx  saq1_1 saq1_2  saq2_1 saq2_2  saq3_1 saq3_2  saq4_1 saq4_2  \
0     27127       3      4       3      4       4      5       1          
1     27128       4      5       4      5       4      5       3      4   
2     27129       4      6       4      5       4      7       2      4   
3     27130       3      4       3      4       3      3       1          
4     27131       4      4       4      4       4      4       4      4   
...     ...     ...    ...     ...    ...     ...    ...     ...    ...   
8117  35244       2      2       3      3       2      2       2      1   
8118  35245       5      6       5      6       4      6       3      3   
8119  35246       3      3       3      3       3      5       3      3   
8120  35247       3      5       3      4       4      5       4      4   
8121  35248       3      5       3      5       3      4       3      4   

      saq5_1  ... bq25  bq26 bq26_1  bq27_1 bq27_2  bq28 bq29  bq30_1 bq30_2  \
0          1  ...   27     2     문과       1      1     1   40    2300   2000   
1          4  ...   58     4    한약학       2      5     1   40                  
2          1  ...   38     4  신문방송학       1      1     1   48    4700   3400   
3          2  ...   39     5     임학       1      1     1   40    4900   4500   
4          4  ...   41     2    인문계       1      1     1   40    3200   2400   
...      ...  ...  ...   ...    ...     ...    ...   ...  ...     ...    ...   
8117       3  ...   34     4   제어계측       1      1     1   40    4200   3200   
8118       3  ...   32     4   일어일문       1      1     1   40    2600   2200   
8119       4  ...   37     4   전기공학       1      1     1   45    4800   4100   
8120       3  ...   38     4     화학       1      1     1   40    5000   4000   
8121       2  ...   47     4  에너지공학       1      1     1   40    5000   4000   

      bq30_3  
0             
1       5800  
2             
3             
4             
...      ...  
8117          
8118          
8119          
8120          
8121          

[8122 rows x 184 columns]

In [5]:
train_2020 = pd.read_csv('know_2020_test_enutf.csv')

## 이상치 파악

In [6]:
test_2020.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8122 entries, 0 to 8121
Data columns (total 184 columns):
 #    Column   Dtype 
---   ------   ----- 
 0    idx      int64 
 1    saq1_1   int64 
 2    saq1_2   object
 3    saq2_1   int64 
 4    saq2_2   object
 5    saq3_1   int64 
 6    saq3_2   object
 7    saq4_1   int64 
 8    saq4_2   object
 9    saq5_1   int64 
 10   saq5_2   object
 11   saq6_1   int64 
 12   saq6_2   object
 13   saq7_1   int64 
 14   saq7_2   object
 15   saq8_1   int64 
 16   saq8_2   object
 17   saq9_1   int64 
 18   saq9_2   object
 19   saq10_1  int64 
 20   saq10_2  object
 21   saq11_1  int64 
 22   saq11_2  object
 23   saq12_1  int64 
 24   saq12_2  object
 25   saq13_1  int64 
 26   saq13_2  object
 27   saq14_1  int64 
 28   saq14_2  object
 29   saq15_1  int64 
 30   saq15_2  object
 31   saq16_1  int64 
 32   saq16_2  object
 33   saq17_1  int64 
 34   saq17_2  object
 35   saq18_1  int64 
 36   saq18_2  object
 37   saq19_1  int64 
 38   saq19_

In [7]:
test_2020.isnull().sum()[:50]

idx        0
saq1_1     0
saq1_2     0
saq2_1     0
saq2_2     0
saq3_1     0
saq3_2     0
saq4_1     0
saq4_2     0
saq5_1     0
saq5_2     0
saq6_1     0
saq6_2     0
saq7_1     0
saq7_2     0
saq8_1     0
saq8_2     0
saq9_1     0
saq9_2     0
saq10_1    0
saq10_2    0
saq11_1    0
saq11_2    0
saq12_1    0
saq12_2    0
saq13_1    0
saq13_2    0
saq14_1    0
saq14_2    0
saq15_1    0
saq15_2    0
saq16_1    0
saq16_2    0
saq17_1    0
saq17_2    0
saq18_1    0
saq18_2    0
saq19_1    0
saq19_2    0
saq20_1    0
saq20_2    0
saq21_1    0
saq21_2    0
saq22_1    0
saq22_2    0
saq23_1    0
saq23_2    0
saq24_1    0
saq24_2    0
saq25_1    0
dtype: int64

In [8]:
test_2020.isnull().sum()[50:100]

saq25_2    0
saq26_1    0
saq26_2    0
saq27_1    0
saq27_2    0
saq28_1    0
saq28_2    0
saq29_1    0
saq29_2    0
saq30_1    0
saq30_2    0
saq31_1    0
saq31_2    0
saq32_1    0
saq32_2    0
saq33_1    0
saq33_2    0
saq34_1    0
saq34_2    0
saq35_1    0
saq35_2    0
saq36_1    0
saq36_2    0
saq37_1    0
saq37_2    0
saq38_1    0
saq38_2    0
saq39_1    0
saq39_2    0
saq40_1    0
saq40_2    0
saq41_1    0
saq41_2    0
saq42_1    0
saq42_2    0
saq43_1    0
saq43_2    0
saq44_1    0
saq44_2    0
vq1        0
vq2        0
vq3        0
vq4        0
vq5        0
vq6        0
vq7        0
vq8        0
vq9        0
vq10       0
vq11       0
dtype: int64

In [9]:
test_2020.isnull().sum()[100:150]

vq12       0
vq13       0
bq1        0
bq2        0
bq3        0
bq4        0
bq4_1a     0
bq4_1b     0
bq4_1c     0
bq5        0
bq5_1      0
bq5_2      0
bq6        0
bq7        0
bq8_1      0
bq8_2      0
bq8_3      0
bq9        0
bq10       0
bq11       0
bq12_1     0
bq12_2     0
bq12_3     0
bq12_4     0
bq12_5     0
bq13_1     0
bq13_2     0
bq13_3     0
bq14_1     0
bq14_2     0
bq14_3     0
bq14_4     0
bq14_5     0
bq14_6     0
bq14_7     0
bq15       0
bq16_1     0
bq16_2     0
bq16_3     0
bq16_4     0
bq16_5     0
bq16_6     0
bq16_7     0
bq16_8     0
bq16_9     0
bq16_10    0
bq17       0
bq18_1     0
bq18_2     0
bq18_3     0
dtype: int64

In [10]:
test_2020.isnull().sum()[150:]

bq18_4     0
bq18_5     0
bq18_6     0
bq18_7     0
bq18_8     0
bq18_9     0
bq18_10    3
bq19       0
bq20       0
bq20_1     0
bq21_1     0
bq21_2     0
bq21_3     0
bq21_4     0
bq22_1     0
bq22_2     0
bq22_3     0
bq22_4     0
bq22_5     0
bq22_6     0
bq23_1     0
bq23_2     0
bq23_3     0
bq24       0
bq25       0
bq26       0
bq26_1     0
bq27_1     0
bq27_2     0
bq28       0
bq29       0
bq30_1     0
bq30_2     0
bq30_3     0
dtype: int64

In [11]:
test_2020[test_2020.bq18_10.isnull()==True].idx

3605    30732
4114    31241
4310    31437
Name: idx, dtype: int64

In [12]:
for i in test_2020.columns : 
    print(i, len(test_2020[i].value_counts()))

idx 8122
saq1_1 5
saq1_2 8
saq2_1 5
saq2_2 8
saq3_1 5
saq3_2 8
saq4_1 5
saq4_2 8
saq5_1 5
saq5_2 8
saq6_1 5
saq6_2 8
saq7_1 5
saq7_2 8
saq8_1 5
saq8_2 8
saq9_1 5
saq9_2 8
saq10_1 5
saq10_2 8
saq11_1 5
saq11_2 8
saq12_1 5
saq12_2 8
saq13_1 5
saq13_2 8
saq14_1 5
saq14_2 8
saq15_1 5
saq15_2 8
saq16_1 5
saq16_2 8
saq17_1 5
saq17_2 8
saq18_1 5
saq18_2 8
saq19_1 5
saq19_2 8
saq20_1 5
saq20_2 8
saq21_1 5
saq21_2 8
saq22_1 5
saq22_2 8
saq23_1 5
saq23_2 8
saq24_1 5
saq24_2 8
saq25_1 5
saq25_2 8
saq26_1 5
saq26_2 8
saq27_1 5
saq27_2 8
saq28_1 5
saq28_2 8
saq29_1 5
saq29_2 8
saq30_1 5
saq30_2 8
saq31_1 5
saq31_2 8
saq32_1 5
saq32_2 8
saq33_1 5
saq33_2 8
saq34_1 5
saq34_2 8
saq35_1 5
saq35_2 8
saq36_1 5
saq36_2 8
saq37_1 5
saq37_2 8
saq38_1 5
saq38_2 8
saq39_1 5
saq39_2 8
saq40_1 5
saq40_2 8
saq41_1 5
saq41_2 8
saq42_1 5
saq42_2 8
saq43_1 5
saq43_2 8
saq44_1 5
saq44_2 8
vq1 5
vq2 5
vq3 5
vq4 5
vq5 5
vq6 5
vq7 5
vq8 5
vq9 5
vq10 5
vq11 5
vq12 5
vq13 5
bq1 21
bq2 5
bq3 6
bq4 2
bq4_1a 1535
bq4_1b 622

In [13]:
test_2020['bq28'].value_counts()

1    7293
2     823
        6
Name: bq28, dtype: int64

1. 결측치

bq18_10    3
idx 30732 31241 31437 -> 주관식이라 drop

2. 이상치
bq2 근무업체규모 5 / 기본값 6

bq28 5 type
3. 주관식 문항 

bq4_1a 1632
bq4_1b 629
bq4_1c 182

bq5_2 훈련사항 1745

bq18_10 직업세계변화 1314

bq20_1 일자리변화 6947

bq26_1 학과 1088

4. 숫자여야하는 col
bq25 나이 60

bq29 근로시간 58
bq30_1 근로소득 329
bq30_2 근로소득 208
bq30_3 근로소득 102
5. 체크해봐야하는 col 

bq22_1 9
bq22_2 9
bq22_3 9
bq22_4 9
bq22_5 9
bq22_6 9
bq23_1 9
bq23_2 9
bq23_3 9

bq27_1 2
bq27_2 7

## 이상치 처리

In [14]:
for col in test_2020:
        test_2020[col].replace(' ', 0, inplace = True)

In [15]:
test_2020.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8122 entries, 0 to 8121
Data columns (total 184 columns):
 #    Column   Dtype 
---   ------   ----- 
 0    idx      int64 
 1    saq1_1   int64 
 2    saq1_2   object
 3    saq2_1   int64 
 4    saq2_2   object
 5    saq3_1   int64 
 6    saq3_2   object
 7    saq4_1   int64 
 8    saq4_2   object
 9    saq5_1   int64 
 10   saq5_2   object
 11   saq6_1   int64 
 12   saq6_2   object
 13   saq7_1   int64 
 14   saq7_2   object
 15   saq8_1   int64 
 16   saq8_2   object
 17   saq9_1   int64 
 18   saq9_2   object
 19   saq10_1  int64 
 20   saq10_2  object
 21   saq11_1  int64 
 22   saq11_2  object
 23   saq12_1  int64 
 24   saq12_2  object
 25   saq13_1  int64 
 26   saq13_2  object
 27   saq14_1  int64 
 28   saq14_2  object
 29   saq15_1  int64 
 30   saq15_2  object
 31   saq16_1  int64 
 32   saq16_2  object
 33   saq17_1  int64 
 34   saq17_2  object
 35   saq18_1  int64 
 36   saq18_2  object
 37   saq19_1  int64 
 38   saq19_

In [16]:
# 주관식문항 

str_li =['bq4_1a','bq4_1b','bq4_1c','bq5_2', 'bq18_10','bq20_1','bq26_1']

In [17]:
for i in test_2020.columns : 
    if i not in str_li : 
        test_2020[i] = test_2020[i].astype(int)
    else : 
        test_2020[i] = test_2020[i].astype(str)

In [18]:
test_2020.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8122 entries, 0 to 8121
Data columns (total 184 columns):
 #    Column   Dtype 
---   ------   ----- 
 0    idx      int32 
 1    saq1_1   int32 
 2    saq1_2   int32 
 3    saq2_1   int32 
 4    saq2_2   int32 
 5    saq3_1   int32 
 6    saq3_2   int32 
 7    saq4_1   int32 
 8    saq4_2   int32 
 9    saq5_1   int32 
 10   saq5_2   int32 
 11   saq6_1   int32 
 12   saq6_2   int32 
 13   saq7_1   int32 
 14   saq7_2   int32 
 15   saq8_1   int32 
 16   saq8_2   int32 
 17   saq9_1   int32 
 18   saq9_2   int32 
 19   saq10_1  int32 
 20   saq10_2  int32 
 21   saq11_1  int32 
 22   saq11_2  int32 
 23   saq12_1  int32 
 24   saq12_2  int32 
 25   saq13_1  int32 
 26   saq13_2  int32 
 27   saq14_1  int32 
 28   saq14_2  int32 
 29   saq15_1  int32 
 30   saq15_2  int32 
 31   saq16_1  int32 
 32   saq16_2  int32 
 33   saq17_1  int32 
 34   saq17_2  int32 
 35   saq18_1  int32 
 36   saq18_2  int32 
 37   saq19_1  int32 
 38   saq19_

In [19]:
## _1에서 1을 선택한 사람중 _2를 답변한 사람이 있는지 확인

for i in range(44) :
    str5 = 'saq' + str(i+1) +'_1'
    str6 = 'saq' + str(i+1) +'_2'
    
    print(test_2020[(test_2020[str5]==1) & (test_2020[str6]!=0)][str6])

Series([], Name: saq1_2, dtype: int32)
Series([], Name: saq2_2, dtype: int32)
Series([], Name: saq3_2, dtype: int32)
Series([], Name: saq4_2, dtype: int32)
Series([], Name: saq5_2, dtype: int32)
Series([], Name: saq6_2, dtype: int32)
Series([], Name: saq7_2, dtype: int32)
Series([], Name: saq8_2, dtype: int32)
Series([], Name: saq9_2, dtype: int32)
Series([], Name: saq10_2, dtype: int32)
Series([], Name: saq11_2, dtype: int32)
Series([], Name: saq12_2, dtype: int32)
Series([], Name: saq13_2, dtype: int32)
Series([], Name: saq14_2, dtype: int32)
Series([], Name: saq15_2, dtype: int32)
Series([], Name: saq16_2, dtype: int32)
Series([], Name: saq17_2, dtype: int32)
Series([], Name: saq18_2, dtype: int32)
Series([], Name: saq19_2, dtype: int32)
Series([], Name: saq20_2, dtype: int32)
Series([], Name: saq21_2, dtype: int32)
Series([], Name: saq22_2, dtype: int32)
Series([], Name: saq23_2, dtype: int32)
Series([], Name: saq24_2, dtype: int32)
Series([], Name: saq25_2, dtype: int32)
Series([]

In [20]:
## 결측치파악(답변있어야하는 셀)
for i in range(44) :
    str5 = 'saq' + str(i+1) +'_1'
    str6 = 'saq' + str(i+1) +'_2'
    
    print(test_2020[(test_2020[str5].isin([2,3,4,5])) & (test_2020[str6]==0)][str6])
    
#saq40_2

Series([], Name: saq1_2, dtype: int32)
Series([], Name: saq2_2, dtype: int32)
Series([], Name: saq3_2, dtype: int32)
Series([], Name: saq4_2, dtype: int32)
Series([], Name: saq5_2, dtype: int32)
Series([], Name: saq6_2, dtype: int32)
Series([], Name: saq7_2, dtype: int32)
Series([], Name: saq8_2, dtype: int32)
Series([], Name: saq9_2, dtype: int32)
Series([], Name: saq10_2, dtype: int32)
Series([], Name: saq11_2, dtype: int32)
Series([], Name: saq12_2, dtype: int32)
Series([], Name: saq13_2, dtype: int32)
Series([], Name: saq14_2, dtype: int32)
Series([], Name: saq15_2, dtype: int32)
Series([], Name: saq16_2, dtype: int32)
Series([], Name: saq17_2, dtype: int32)
Series([], Name: saq18_2, dtype: int32)
Series([], Name: saq19_2, dtype: int32)
Series([], Name: saq20_2, dtype: int32)
Series([], Name: saq21_2, dtype: int32)
Series([], Name: saq22_2, dtype: int32)
Series([], Name: saq23_2, dtype: int32)
Series([], Name: saq24_2, dtype: int32)
Series([], Name: saq25_2, dtype: int32)
Series([]

In [21]:
# saq40_2 최빈값 확인

train_2020[(train_2020.saq40_1.isin([2,3,4,5]))]['saq40_2'].value_counts()

4    1941
3    1497
5    1326
2     816
6     301
1     122
7      32
Name: saq40_2, dtype: int64

In [22]:
test_2020.iloc[7136,80]=4

In [23]:
# 값이 잘 변경되었는지 확인 
test_2020[(test_2020.saq40_1.isin([2,3,4,5]))]['saq40_2'].value_counts()

4    1941
3    1497
5    1326
2     816
6     301
1     122
7      32
Name: saq40_2, dtype: int64

1. saq40_2 
train saq40_1에서 2,3,4,5를 선택한 사람들의 saq40_2 답변 중 최빈값을 넣기 -> 4로 처리 


2. 주관식 항목 drop예정으로 확인 안함

bq18_10 bq20_1

3. 학과
bq26_1 빈셀 없음

4. bq27_1 ~ bq30_3 
bq28,29 는 조건에 맞게 대답하려면 빈칸이 많아야하는데 그렇지 않음 그냥 빈셀을 채워주는 식으로 처리 할 예정
bq28 빈셀 6개 / bq27_1 :1 2개, 2 4개

28189 28916 31841 34587 35066 -> 2

33528 -> 1


bq29 빈셀 없음

2017과 동일하게 처리 bq30_1 빈셀 없음 bq30_2 빈셀 없음 bq30_3 빈셀 없음

In [252]:
test_2020.iloc[6401, 179] =1

In [253]:
test_2020.iloc[[1062, 1789, 4714,7460, 7939], 179] =2

In [254]:
## 자격증 항목 없음, 없다 처리 


test_2020['bq4_1b'].replace(0, '0', inplace = True)


for i in test_2020.bq4_1b:
    if '없' in i : 
        print(i)

        
test_2020['bq4_1b'].replace('없다', '0', inplace = True)

없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다


In [255]:
for i in test_2020.bq4_1b:
    if '없' in i :
        print(i)

In [256]:
test_2020.to_csv("know_2020_test_enutf.csv", encoding='utf8', index=False)
test_2020.to_csv("know_2020_test_encp.csv", encoding='cp949', index=False)

In [257]:
df_utf = pd.read_csv("know_2020_test_enutf.csv")
df_utf.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8122 entries, 0 to 8121
Data columns (total 184 columns):
 #    Column   Dtype 
---   ------   ----- 
 0    idx      int64 
 1    saq1_1   int64 
 2    saq1_2   int64 
 3    saq2_1   int64 
 4    saq2_2   int64 
 5    saq3_1   int64 
 6    saq3_2   int64 
 7    saq4_1   int64 
 8    saq4_2   int64 
 9    saq5_1   int64 
 10   saq5_2   int64 
 11   saq6_1   int64 
 12   saq6_2   int64 
 13   saq7_1   int64 
 14   saq7_2   int64 
 15   saq8_1   int64 
 16   saq8_2   int64 
 17   saq9_1   int64 
 18   saq9_2   int64 
 19   saq10_1  int64 
 20   saq10_2  int64 
 21   saq11_1  int64 
 22   saq11_2  int64 
 23   saq12_1  int64 
 24   saq12_2  int64 
 25   saq13_1  int64 
 26   saq13_2  int64 
 27   saq14_1  int64 
 28   saq14_2  int64 
 29   saq15_1  int64 
 30   saq15_2  int64 
 31   saq16_1  int64 
 32   saq16_2  int64 
 33   saq17_1  int64 
 34   saq17_2  int64 
 35   saq18_1  int64 
 36   saq18_2  int64 
 37   saq19_1  int64 
 38   saq19_

In [258]:
df_cp = pd.read_csv("know_2020_test_encp.csv", encoding='cp949')
df_cp.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8122 entries, 0 to 8121
Data columns (total 184 columns):
 #    Column   Dtype 
---   ------   ----- 
 0    idx      int64 
 1    saq1_1   int64 
 2    saq1_2   int64 
 3    saq2_1   int64 
 4    saq2_2   int64 
 5    saq3_1   int64 
 6    saq3_2   int64 
 7    saq4_1   int64 
 8    saq4_2   int64 
 9    saq5_1   int64 
 10   saq5_2   int64 
 11   saq6_1   int64 
 12   saq6_2   int64 
 13   saq7_1   int64 
 14   saq7_2   int64 
 15   saq8_1   int64 
 16   saq8_2   int64 
 17   saq9_1   int64 
 18   saq9_2   int64 
 19   saq10_1  int64 
 20   saq10_2  int64 
 21   saq11_1  int64 
 22   saq11_2  int64 
 23   saq12_1  int64 
 24   saq12_2  int64 
 25   saq13_1  int64 
 26   saq13_2  int64 
 27   saq14_1  int64 
 28   saq14_2  int64 
 29   saq15_1  int64 
 30   saq15_2  int64 
 31   saq16_1  int64 
 32   saq16_2  int64 
 33   saq17_1  int64 
 34   saq17_2  int64 
 35   saq18_1  int64 
 36   saq18_2  int64 
 37   saq19_1  int64 
 38   saq19_